### PR TITLE
DAgger

### PR DESCRIPTION
**Pull request by qxcv**
This commit adds a DAgger implementation on top of the existing BC implementation. I've been using an interactive demo collection workflow, so some of the design choices might not work as well with automated demonstrators (e.g. I use a plain gym `env` instead of a `vecenv` for data collection because a human demonstrator can't be vectorised :smile:). Still, worth having this for now & we can revisit if it becomes a problem.

Also this is based on the `save-bc-pols` branch, so #125 should be merged first.

### CODE REVIEW

File: src/imitation/algorithms/dagger.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+    if round_num is None:
+      round_num = self.round_num
+    return os.path.join(self.scratch_dir, 'demos', f'round-{round_num:03d}')
+
+  def _check_has_latest_demos(self):
+    demo_dir = self._demo_dir_path_for_round()
+    demo_paths = self._get_demo_paths(demo_dir) \
+        if os.path.isdir(demo_dir) else []
+    if len(demo_paths) == 0:
+      raise NeedDemosException(
```
Comment (shwang) [inline src/imitation/algorithms/dagger.py]:
```suggestion
      raise NeedsDemosException(
```

File: src/imitation/algorithms/dagger.py [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+
+  def schedule(i: int) -> float:
+    assert i >= 0
+    return min(1, max(0, (rampdown_rounds - i) / rampdown_rounds))
+
+  return schedule
+
+
+def _save_trajectory(npz_path, trajectory):
+  """Save a trajectory as a compressed Numpy file."""
```
[the remark this answers is not in the document]
Comment (shwang) [reply src/imitation/algorithms/dagger.py]: Oh, it looks like this save format has to do with how we are collecting and saving new demonstrations at the same time?

[the remark this answers is not in the document]
Comment (qxcv) [reply src/imitation/algorithms/dagger.py]: Yes, the point is to save collected trajectories incrementally, so it doesn't matter if the collection process crashes or is terminated by the user. Also this strategy allows trajectories to be manually deleted.

**Comment by shwang** [comment]
@qxcv Could you merge this with master? I tried myself and  in some conflicts didn't know whether to use HEAD or master.

**Comment by qxcv** [comment]
I just merged, although there are still some tests failing for mysterious reasons (maybe policy serialisation issues?). e.g. one fails with `AttributeError: 'DummyVecEnv' object has no attribute 'normalize_obs'` when doing an expert rollout. Will finally sort out this (and my other PR) tomorrow.

File: src/imitation/util/rollout.py [hunk 3]
```diff
@@ -79,24 +80,29 @@ class Transitions(NamedTuple):
...
 
 
-class _TrajectoryAccumulator:
-  """Accumulates trajectories step-by-step.
-
-  Used in `generate_trajectories()` only, for collecting completed trajectories
-  while ignoring partially-completed trajectories.
-  """
+class TrajectoryAccumulator:
+  """Accumulates trajectories step-by-step. Useful for collecting completed
```
[the remark this answers is not in the document]
Comment (qxcv) [reply src/imitation/util/rollout.py]: Like this

```python
"""
words
"""
```

instead of this

```python
"""words"""
```

?

**Comment by qxcv** [comment]
Comments should be addressed now :+1:

**Pull request merged**